pkg/investigations/chgm/chgm_hibernation_check.go

```diff
@@ -0,0 +1,87 @@
+package chgm
+
+import (
+	"sort"
+	"time"
+
+	v1 "github.com/openshift-online/ocm-sdk-go/clustersmgmt/v1"
+	servicelogsv1 "github.com/openshift-online/ocm-sdk-go/servicelogs/v1"
+	"github.com/openshift/configuration-anomaly-detection/pkg/ocm"
+)
+
+const RECENT_WAKEUP_TIME = 2 * time.Hour
+
+// 30 Days is always a problem as kubelet certificates will be expired
+const HIBERNATION_TOO_LONG = 30 * 24 * time.Hour
+const HIBERNATION_START_EVENT = "cluster_state_hibernating"
+const HIBERNATION_END_EVENT = "cluster_state_ready"
+const HIBERNATION_ONGOING_EVENT = "cluster_state_hibernating"
+const HIBERNATION_RESUME_EVENT = "cluster_state_resuming"
+
+type HibernationPeriod struct {
+	HibernationDuration time.Duration
+	DehibernationTime   time.Time
+}
+
+func HibernatedTooLong(hibernations []*HibernationPeriod, now time.Time) bool {
+	if len(hibernations) == 0 {
+		return false
+	}
+	latestHibernation := hibernations[len(hibernations)-1]
+	// The cluster was woken up within the RECENT_WAKEUP_TIME which might
+	// indicate a CSR problem.
+	if now.Sub(latestHibernation.DehibernationTime) < RECENT_WAKEUP_TIME {
+		return false
+	}
+	// Only clusters that have hibernated for a long time are susceptible to
+	// have cert issues.
+	if latestHibernation.HibernationDuration >= HIBERNATION_TOO_LONG {
+		return true
+	}
+	return false
+}
+
+func getHibernationStatusForCluster(ocmClient ocm.Client, cluster *v1.Cluster) ([]*HibernationPeriod, error) {
+	filter := "log_type='cluster-state-updates'"
+	clusterStateUpdates, err := ocmClient.GetServiceLog(cluster, filter)
+	if err != nil {
+		return nil, err
+	}
+	return CreateHibernationTimeLine(clusterStateUpdates.Items().Slice()), nil
+}
+
+func CreateHibernationTimeLine(clusterStateUpdates []*servicelogsv1.LogEntry) []*HibernationPeriod {
```

Comment:
**Question:** same question about it being public :D

Reply:
I kind of forgot that all of those are in the same package, so I made all of them private now.